// xapi-node/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProfit = exports.getContractValue = exports.Counter = exports.Logger = exports.TradeRecord = exports.LimitPosition = exports.OpenPosition = exports.Timer = exports.Time = exports.parseJsonLogin = exports.XAPI = exports.errorCode = exports.Candle = exports.CHART_RATE_LIMIT_BY_PERIOD = exports.REQUEST_STATUS_FIELD = exports.STATE_FIELD = exports.TYPE_FIELD = exports.PERIOD_FIELD = exports.DAY_FIELD = exports.CMD_FIELD = exports.XAPIv2 = void 0;
const XAPI_1 = require("./v2/core/XAPI");
Object.defineProperty(exports, "XAPI", { enumerable: true, get: function () { return XAPI_1.XAPI; } });
const XAPI_2 = require("./v2/core/XAPI");
Object.defineProperty(exports, "XAPIv2", { enumerable: true, get: function () { return XAPI_2.XAPI; } });
const Enum_1 = require("./v2/interface/Enum");
Object.defineProperty(exports, "Candle", { enumerable: true, get: function () { return Enum_1.Candle; } });
Object.defineProperty(exports, "CHART_RATE_LIMIT_BY_PERIOD", { enumerable: true, get: function () { return Enum_1.CHART_RATE_LIMIT_BY_PERIOD; } });
Object.defineProperty(exports, "CMD_FIELD", { enumerable: true, get: function () { return Enum_1.CMD_FIELD; } });
Object.defineProperty(exports, "DAY_FIELD", { enumerable: true, get: function () { return Enum_1.DAY_FIELD; } });
Object.defineProperty(exports, "errorCode", { enumerable: true, get: function () { return Enum_1.errorCode; } });
Object.defineProperty(exports, "PERIOD_FIELD", { enumerable: true, get: function () { return Enum_1.PERIOD_FIELD; } });
Object.defineProperty(exports, "REQUEST_STATUS_FIELD", { enumerable: true, get: function () { return Enum_1.REQUEST_STATUS_FIELD; } });
Object.defineProperty(exports, "STATE_FIELD", { enumerable: true, get: function () { return Enum_1.STATE_FIELD; } });
Object.defineProperty(exports, "TYPE_FIELD", { enumerable: true, get: function () { return Enum_1.TYPE_FIELD; } });
const parseJsonLogin_1 = require("./v2/utils/parseJsonLogin");
Object.defineProperty(exports, "parseJsonLogin", { enumerable: true, get: function () { return parseJsonLogin_1.parseJsonLogin; } });
const Time_1 = require("./v2/utils/Time");
Object.defineProperty(exports, "Time", { enumerable: true, get: function () { return Time_1.Time; } });
const Timer_1 = require("./v2/utils/Timer");
Object.defineProperty(exports, "Timer", { enumerable: true, get: function () { return Timer_1.Timer; } });
const OpenPosition_1 = require("./v2/core/Trading/OpenPosition");
Object.defineProperty(exports, "OpenPosition", { enumerable: true, get: function () { return OpenPosition_1.OpenPosition; } });
const LimitPosition_1 = require("./v2/core/Trading/LimitPosition");
Object.defineProperty(exports, "LimitPosition", { enumerable: true, get: function () { return LimitPosition_1.LimitPosition; } });
const TradeRecord_1 = require("./v2/core/TradeRecord");
Object.defineProperty(exports, "TradeRecord", { enumerable: true, get: function () { return TradeRecord_1.TradeRecord; } });
const Logger_1 = require("./v2/utils/Logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return Logger_1.Logger; } });
const Counter_1 = require("./v2/utils/Counter");
Object.defineProperty(exports, "Counter", { enumerable: true, get: function () { return Counter_1.Counter; } });
exports.default = XAPI_1.XAPI;
function getContractValue({ price, lot, contractSize, currency, currencyProfit, }) {
    return lot * contractSize * (currency === currencyProfit ? price : 1);
}
exports.getContractValue = getContractValue;
function getProfit({ openPrice, closePrice, isBuy, lot, contractSize, }) {
    return (isBuy ? closePrice - openPrice : openPrice - closePrice) * lot * contractSize;
}
exports.getProfit = getProfit;
