// xapi-node/build/v2/core/TradeRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TradeRecord = void 0;
const Enum_1 = require("../interface/Enum");
const getPositionType_1 = require("../utils/getPositionType");
class TradeRecord {
    constructor(params) {
        this.close_time = params.close_time;
        this.closed = params.closed;
        this.cmd = params.cmd;
        this.comment = params.comment;
        this.commission = params.commission;
        this.customComment = params.customComment;
        this.digits = params.digits;
        this.expiration = params.expiration;
        this.margin_rate = params.margin_rate;
        this.offset = params.offset;
        this.open_price = params.open_price;
        this.open_time = params.open_time;
        this.order = params.order;
        this.order2 = params.order2;
        this.position = params.position;
        this.sl = params.sl;
        this.storage = params.storage;
        this.symbol = params.symbol;
        this.tp = params.tp;
        this.volume = params.volume;
        this.close_price = this.position_type === Enum_1.PositionType.closed ? params.close_price : undefined;
    }
    get position_type() {
        return (0, getPositionType_1.getPositionType)({ cmd: this.cmd, closed: this.closed, close_time: this.close_time });
    }
    valueOf() {
        return {
            close_time: this.close_time,
            close_price: this.close_price,
            closed: this.closed,
            cmd: this.cmd,
            comment: this.comment,
            commission: this.commission,
            customComment: this.customComment,
            digits: this.digits,
            expiration: this.expiration,
            margin_rate: this.margin_rate,
            offset: this.offset,
            open_price: this.open_price,
            open_time: this.open_time,
            order: this.order,
            order2: this.order2,
            position: this.position,
            sl: this.sl,
            storage: this.storage,
            symbol: this.symbol,
            tp: this.tp,
            volume: this.volume,
        };
    }
}
exports.TradeRecord = TradeRecord;
