// xapi-node/build/v2/core/Trading/LimitPosition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimitPosition = void 0;
const TradeRecord_1 = require("../TradeRecord");
const Enum_1 = require("../../interface/Enum");
class LimitPosition extends TradeRecord_1.TradeRecord {
    constructor(XAPI, params) {
        super(params);
        this.XAPI = XAPI;
    }
    close(customComment = '') {
        return this.XAPI.trading.tradeTransaction({
            order: this.order,
            symbol: this.symbol,
            type: Enum_1.TYPE_FIELD.DELETE,
            customComment: customComment || undefined,
        });
    }
    modify(params) {
        return this.XAPI.trading.tradeTransaction({
            order: this.order,
            type: Enum_1.TYPE_FIELD.MODIFY,
            tp: params.tp,
            sl: params.sl,
            price: params.price,
            expiration: params.expiration,
            customComment: params.customComment,
        });
    }
}
exports.LimitPosition = LimitPosition;
