// xapi-node/build/v2/core/Trading/OpenPosition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenPosition = void 0;
const TradeRecord_1 = require("../TradeRecord");
const Enum_1 = require("../../interface/Enum");
class OpenPosition extends TradeRecord_1.TradeRecord {
    constructor(XAPI, params) {
        super(params);
        this.XAPI = XAPI;
    }
    close(volume = undefined, customComment = '') {
        return this.XAPI.trading.tradeTransaction({
            order: this.order,
            type: Enum_1.TYPE_FIELD.CLOSE,
            volume: volume === undefined ? this.volume : volume,
            symbol: this.symbol,
            price: 1,
            customComment: customComment || undefined,
        });
    }
    modify(params) {
        return this.XAPI.trading.tradeTransaction({
            order: this.order,
            type: Enum_1.TYPE_FIELD.MODIFY,
            tp: params.tp,
            sl: params.sl,
            offset: params.offset,
            customComment: params.customComment || undefined,
        });
    }
}
exports.OpenPosition = OpenPosition;
