// xapi-node/build/v2/interface/Enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Currency2Pair = exports.errorCode = exports.Candle = exports.PositionType = exports.CHART_RATE_LIMIT_BY_PERIOD = exports.PERIOD_FIELD = exports.STATE_FIELD = exports.TYPE_FIELD = exports.CMD_FIELD = exports.DAY_FIELD = exports.REQUEST_STATUS_FIELD = void 0;
var REQUEST_STATUS_FIELD;
(function (REQUEST_STATUS_FIELD) {
    REQUEST_STATUS_FIELD[REQUEST_STATUS_FIELD["ERROR"] = 0] = "ERROR";
    REQUEST_STATUS_FIELD[REQUEST_STATUS_FIELD["PENDING"] = 1] = "PENDING";
    REQUEST_STATUS_FIELD[REQUEST_STATUS_FIELD["ACCEPTED"] = 3] = "ACCEPTED";
    REQUEST_STATUS_FIELD[REQUEST_STATUS_FIELD["REJECTED"] = 4] = "REJECTED";
})(REQUEST_STATUS_FIELD || (exports.REQUEST_STATUS_FIELD = REQUEST_STATUS_FIELD = {}));
var DAY_FIELD;
(function (DAY_FIELD) {
    DAY_FIELD[DAY_FIELD["MONDAY"] = 1] = "MONDAY";
    DAY_FIELD[DAY_FIELD["TUESDAY"] = 2] = "TUESDAY";
    DAY_FIELD[DAY_FIELD["WEDNESDAY"] = 3] = "WEDNESDAY";
    DAY_FIELD[DAY_FIELD["THURSDAY"] = 4] = "THURSDAY";
    DAY_FIELD[DAY_FIELD["FRIDAY"] = 5] = "FRIDAY";
    DAY_FIELD[DAY_FIELD["SATURDAY"] = 6] = "SATURDAY";
    DAY_FIELD[DAY_FIELD["SUNDAY"] = 7] = "SUNDAY";
})(DAY_FIELD || (exports.DAY_FIELD = DAY_FIELD = {}));
var CMD_FIELD;
(function (CMD_FIELD) {
    CMD_FIELD[CMD_FIELD["BUY"] = 0] = "BUY";
    CMD_FIELD[CMD_FIELD["SELL"] = 1] = "SELL";
    CMD_FIELD[CMD_FIELD["BUY_LIMIT"] = 2] = "BUY_LIMIT";
    CMD_FIELD[CMD_FIELD["SELL_LIMIT"] = 3] = "SELL_LIMIT";
    CMD_FIELD[CMD_FIELD["BUY_STOP"] = 4] = "BUY_STOP";
    CMD_FIELD[CMD_FIELD["SELL_STOP"] = 5] = "SELL_STOP";
    CMD_FIELD[CMD_FIELD["BALANCE"] = 6] = "BALANCE";
    CMD_FIELD[CMD_FIELD["CREDIT"] = 7] = "CREDIT";
})(CMD_FIELD || (exports.CMD_FIELD = CMD_FIELD = {}));
var TYPE_FIELD;
(function (TYPE_FIELD) {
    TYPE_FIELD[TYPE_FIELD["OPEN"] = 0] = "OPEN";
    TYPE_FIELD[TYPE_FIELD["PENDING"] = 1] = "PENDING";
    TYPE_FIELD[TYPE_FIELD["CLOSE"] = 2] = "CLOSE";
    TYPE_FIELD[TYPE_FIELD["MODIFY"] = 3] = "MODIFY";
    TYPE_FIELD[TYPE_FIELD["DELETE"] = 4] = "DELETE";
})(TYPE_FIELD || (exports.TYPE_FIELD = TYPE_FIELD = {}));
var STATE_FIELD;
(function (STATE_FIELD) {
    STATE_FIELD["MODIFIED"] = "Modified";
    STATE_FIELD["DELETED"] = "Deleted";
})(STATE_FIELD || (exports.STATE_FIELD = STATE_FIELD = {}));
var PERIOD_FIELD;
(function (PERIOD_FIELD) {
    PERIOD_FIELD[PERIOD_FIELD["PERIOD_M1"] = 1] = "PERIOD_M1";
    PERIOD_FIELD[PERIOD_FIELD["PERIOD_M5"] = 5] = "PERIOD_M5";
    PERIOD_FIELD[PERIOD_FIELD["PERIOD_M15"] = 15] = "PERIOD_M15";
    PERIOD_FIELD[PERIOD_FIELD["PERIOD_M30"] = 30] = "PERIOD_M30";
    PERIOD_FIELD[PERIOD_FIELD["PERIOD_H1"] = 60] = "PERIOD_H1";
    PERIOD_FIELD[PERIOD_FIELD["PERIOD_H4"] = 240] = "PERIOD_H4";
    PERIOD_FIELD[PERIOD_FIELD["PERIOD_D1"] = 1440] = "PERIOD_D1";
    PERIOD_FIELD[PERIOD_FIELD["PERIOD_W1"] = 10080] = "PERIOD_W1";
    PERIOD_FIELD[PERIOD_FIELD["PERIOD_MN1"] = 43200] = "PERIOD_MN1";
})(PERIOD_FIELD || (exports.PERIOD_FIELD = PERIOD_FIELD = {}));
exports.CHART_RATE_LIMIT_BY_PERIOD = {
    PERIOD_M1: 28800,
    PERIOD_M5: 17280,
    PERIOD_M15: 5760,
    PERIOD_M30: 6720,
    PERIOD_H1: 3360,
    PERIOD_H4: 1560,
    PERIOD_D1: 19200,
    PERIOD_W1: 3840,
    PERIOD_MN1: 960,
};
var PositionType;
(function (PositionType) {
    PositionType[PositionType["open"] = 0] = "open";
    PositionType[PositionType["closed"] = 1] = "closed";
    PositionType[PositionType["limit"] = 2] = "limit";
    PositionType[PositionType["source"] = 3] = "source";
})(PositionType || (exports.PositionType = PositionType = {}));
var Candle;
(function (Candle) {
    Candle[Candle["timestamp"] = 0] = "timestamp";
    Candle[Candle["open"] = 1] = "open";
    Candle[Candle["high"] = 2] = "high";
    Candle[Candle["low"] = 3] = "low";
    Candle[Candle["close"] = 4] = "close";
    Candle[Candle["volume"] = 5] = "volume";
})(Candle || (exports.Candle = Candle = {}));
var errorCode;
(function (errorCode) {
    errorCode["XAPINODE_0"] = "XAPINODE_0";
    errorCode["XAPINODE_1"] = "XAPINODE_1";
    errorCode["XAPINODE_2"] = "XAPINODE_2";
    errorCode["XAPINODE_3"] = "XAPINODE_3";
    errorCode["XAPINODE_4"] = "XAPINODE_4";
    errorCode["XAPINODE_BE103"] = "XAPINODE_BE103";
    errorCode["BE005"] = "BE005";
    errorCode["BE118"] = "BE118";
})(errorCode || (exports.errorCode = errorCode = {}));
exports.Currency2Pair = {
    HUF: 'EURHUF',
    USD: 'EURUSD',
    JPY: 'USDJPY',
    GBP: 'EURGBP',
    TRY: 'EURTRY',
    CHF: 'USDCHF',
    CZK: 'USDCZK',
    BRL: 'USDBRL',
    PLN: 'USDPLN',
    MXN: 'USDMXN',
    ZAR: 'USDZAR',
    RON: 'USDRON',
    AUD: 'EURAUD',
    CAD: 'USDCAD',
    SEK: 'USDSEK',
    NOK: 'EURNOK',
    NZD: 'EURNZD',
    EUR: 'DE30',
    CLP: 'USDCLP',
    DKK: 'VWS.DK_4',
    BTC: 'XEMBTC',
    ETH: 'TRXETH',
};
